lua/barbar/groups.lua

```diff
@@ -0,0 +1,206 @@
+--
+-- groups.lua
+--
+
+local table_insert = table.insert
+local strcharpart = vim.fn.strcharpart --- @type function
+local strwidth = vim.api.nvim_strwidth --- @type function
+
+local utils = require'barbar.utils'
+
+local m = {}
+
+--- Sums the width of the groups
+--- @param groups barbar.render.group[]
+--- @return integer
+function m.width(groups)
+  local result = 0
+  for _, group in ipairs(groups) do
+    result = result + strwidth(group.text)
+  end
+  return result
+end
+
+--- Concatenates some `groups` into a valid tabline string.
+--- @param groups barbar.render.group[]
+--- @return string
+function m.to_string(groups)
+  local result = ''
+
+  for _, group in ipairs(groups) do
+    -- NOTE: We have to escape the text in case it contains '%', which is a special character to the
+    --       tabline.
+    --       To escape '%', we make it '%%'. It just so happens that '%' is also a special character
+    --       in Lua, so we have write '%%' to mean '%'.
+    result = result .. group.hl .. group.text:gsub('%%', '%%%%')
+  end
+
+  return result
+end
+
+--- Concatenates some `groups` into a raw string.
+--- For debugging purposes.
+--- @param groups barbar.render.group[]
+--- @return string
+--- @diagnostic disable-next-line:unused-function,unused-local
+function m.to_raw_string(groups)
+  local result = ''
+
+  for _, group in ipairs(groups) do
+    result = result .. group.text
+  end
+
+  return result
+end
+
+--- Insert `other` into `groups` at the `position`.
+--- @param groups barbar.render.group[]
+--- @param position integer
+--- @param group barbar.render.group
+--- @return barbar.render.group[] with_insertions
+function m.insert(groups, position, group)
+  return m.insert_many(groups, position, { group })
+end
+
+--- Insert `others` into `groups` at the `position`.
+--- @param groups barbar.render.group[]
+--- @param position integer
+--- @param others barbar.render.group[]
+--- @return barbar.render.group[] with_insertions
+function m.insert_many(groups, position, others)
+
+  if position < 0 then
+    local others_width = m.width(others)
+    local others_end = position + others_width
+
+    if others_end < 0 then
+      return groups
+    end
+
+    local available_width = others_end
+
+    position = 0
+    others = m.slice_left(others, available_width)
+  end
```

Comment:
We now allow inserting groups before position `0`. We use this for scrolling instead of the previous approach.